File: src/main/GO/leetcode/lists/offerII/Offer28.go
package offerII

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/15 14:02
 */

//https://leetcode.com/problems/flatten-a-multilevel-doubly-linked-list/solutions/2728002/java-4-solutions-iteration-recursive-constant-space/
//指针赋值
func flatten(root *Node) *Node {
	if root == nil {
		return root
	}
	cur := &Node{}
	var dfs func(node *Node)
	dfs = func(node *Node) {
		if node == nil {
			return
		}
		cur.Next = node
		node.Prev = cur
		cur = cur.Next
		last := node.Next
		if node.Child != nil {
			dfs(node.Child)
			node.Child = nil
		}
		if last != nil {
			dfs(last)
		}
	}
	dfs(root)
	root.Prev = nil
	return root
}

File: java/src/leetcode/offer/Offer28.java
package leetcode.offer;

import common.TreeNode;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/9/12 22:00
 */
public class Offer28 {


    public boolean isSymmetric(TreeNode root) {

        if (root == null) {
            return true;
        }
        return isSymmetric(root.right, root.left);
    }

    public boolean isSymmetric(TreeNode right, TreeNode left) {


        if (right == null && left == null) {
            return true;
        }
        if (right == null || left == null) {
            return false;
        }
        if (right.val != left.val) {
            return false;
        }
        return isSymmetric(right.right, left.left) && isSymmetric(right.left, left.right);
    }
}
